File: sdk/node/cpm_sdk.js
/**
 * CPM Node.js SDK — Native JavaScript/TypeScript Client for CPM / UPM Polyglot Bridge RPC
 *
 * Usage:
 *   const { CpmBridge } = require('./cpm_sdk');
 *   const bridge = new CpmBridge();
 *   const result = await bridge.call('python:math.sqrt', [144.0]);
 *   console.log(result);
 */

const { execFile } = require('child_process');
const path = require('path');
const fs = require('fs');

class CpmBridge {
    constructor(cpmBin) {
        this.cpmBin = cpmBin || this._findCpmBin();
    }

    _findCpmBin() {
        let curr = path.resolve(__dirname);
        for (let i = 0; i < 5; i++) {
            for (const rel of [path.join('target', 'release', 'cpm.exe'), path.join('target', 'debug', 'cpm.exe')]) {
                const p = path.join(curr, rel);
                if (fs.existsSync(p)) {
                    return p;
                }
            }
            const parent = path.dirname(curr);
            if (parent === curr) break;
            curr = parent;
        }
        return 'cpm';
    }

    /**
     * Call a foreign language method over CPM stdio RPC bridge.
     * @param {string} target Method target in format 'language:method' (e.g. 'python:math.sqrt')
     * @param {Array} args Arguments array
     * @returns {Promise<any>} Response result from foreign runtime
     */
    call(target, args = []) {
        return new Promise((resolve, reject) => {
            const argsJson = JSON.stringify(args);
            execFile(this.cpmBin, ['bridge', 'call', target, argsJson], { encoding: 'utf8' }, (err, stdout, stderr) => {
                if (err) {
                    return reject(new Error(`CPM Bridge Error: ${stderr.trim() || stdout.trim()}`));
                }

                const lines = stdout.trim().split('\n');
                let capture = false;
                const jsonLines = [];

                for (const line of lines) {
                    if (line.includes('Response received:')) {
                        capture = true;
                        continue;
                    }
                    if (capture && (line.includes('round-trip via stdio RPC') || !line.trim())) {
                        if (jsonLines.length > 0) break;
                    }
                    if (capture) {
                        jsonLines.push(line);
                    }
                }

                const rawJson = jsonLines.join('\n').trim();
                if (rawJson) {
                    try {
                        return resolve(JSON.parse(rawJson));
                    } catch (pErr) {
                        return resolve(rawJson);
                    }
                }
                resolve(stdout.trim());
            });
        });
    }

    /**
     * Inspect registered methods on a foreign language host.
     * @param {string} language Language name (python, node)
     * @returns {Promise<string>} Output method list
     */
    inspect(language) {
        return new Promise((resolve, reject) => {
            execFile(this.cpmBin, ['bridge', 'inspect', language], { encoding: 'utf8' }, (err, stdout, stderr) => {
                if (err) return reject(err);
                resolve(stdout.trim());
            });
        });
    }
}

if (require.main === module) {
    (async () => {
        const bridge = new CpmBridge();
        console.log('Testing CPM Node.js SDK:');
        try {
            const res = await bridge.call('python:math.sqrt', [144.0]);
            console.log('  sqrt(144.0) =', res);
        } catch (e) {
            console.error('  SDK Error:', e.message);
        }
    })();
}

module.exports = { CpmBridge };

File: sdk/dart/cpm_sdk.dart
/// CPM Dart / Flutter SDK — Native Client for CPM / UPM Polyglot Bridge RPC
///
/// Usage:
///     import 'package:cpm_sdk/cpm_sdk.dart';
///     final bridge = CpmBridge();
///     final result = await bridge.call('python:math.sqrt', [144.0]);

import 'dart:convert';
import 'dart:io';

class CpmBridge {
  final String cpmBin;

  CpmBridge({String? cpmBin}) : cpmBin = cpmBin ?? _findCpmBin();

  static String _findCpmBin() {
    var curr = Directory.current;
    for (var i = 0; i < 5; i++) {
      for (var rel in ['target/release/cpm.exe', 'target/debug/cpm.exe', 'target/release/cpm', 'target/debug/cpm']) {
        var f = File('${curr.path}/$rel');
        if (f.existsSync()) {
          return f.path;
        }
      }
      var parent = curr.parent;
      if (parent.path == curr.path) break;
      curr = parent;
    }
    return 'cpm';
  }

  Future<dynamic> call(String target, List<dynamic> args) async {
    final argsJson = jsonEncode(args);
    final res = await Process.run(cpmBin, ['bridge', 'call', target, argsJson]);

    if (res.exitCode != 0) {
      throw Exception('CPM Bridge Call Error: ${res.stderr.toString().trim()}');
    }

    final stdoutStr = res.stdout.toString().trim();
    final lines = stdoutStr.split('\n');
    var capture = false;
    final jsonLines = <String>[];

    for (var line in lines) {
      if (line.contains('Response received:')) {
        capture = true;
        continue;
      }
      if (capture && (line.contains('round-trip via stdio RPC') || line.trim().isEmpty)) {
        if (jsonLines.isNotEmpty) break;
      }
      if (capture) {
        jsonLines.add(line);
      }
    }

    final rawJson = jsonLines.join('\n').trim();
    if (rawJson.isNotEmpty) {
      return jsonDecode(rawJson);
    }
    return stdoutStr;
  }
}
